ENH: Validate and disable CPU features in runtime

seiko2plus opened the pull request:
### This pullrequest changes

---
#### Validate CPU baseline features during the load of NumPy module

Ensure that the baseline features are supported by the running machine, during the load of umath module or while calling C function `npy_cpu_init()` .

Python runtime error will be raised, If one of the baseline features isn't supported. 

---

#### Add ability to disable dispatched CPU features in runtime

Through environment variable `NPY_DISABLE_CPU_FEATURES`, the parsing process happens during the load of umath module or calling C function `npy_cpu_init()` .

An example (unix-like):

```bash
$ export NPY_DISABLE_CPU_FEATURES="FMA3 AVX2 AVX512F AVX512_SKX"
```

**NOTES**:

- CPU features must be in upper-case
- CPU features must be part of dispatched features, disabling CPU baseline features isn't possible.
- space or comma or both can be used as a separator 
- The environment variable must be set before the load of NumPy module, otherwise it will not be affected.
- Any misuse of `NPY_DISABLE_CPU_FEATURES` will cause a Python runtime error.

seiko2plus committed:
ENH: Validate CPU baseline features during the load of NumPy module.

    this validation will cause Python runtime error if the
    NumPy build configured with baseline features that are
    not supported by current hardware.

seiko2plus committed:
ENH: Add ability to disable certain dispatch-able CPU features in runtime

  via an environment variable 'NPY_DISABLE_CPU_FEATURES'.
  Any misuse of 'NPY_DISABLE_CPU_FEATURES' will cause Python runtime error.

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -13,6 +13,39 @@ static unsigned char npy__cpu_have[NPY_CPU_FEATURE_MAX];
+ * Return CPU feature ID if the feature part of NumPy build CPU baseline
+ * otherwise return 0
+*/
+#if !defined(NPY_DISABLE_OPTIMIZATION) && NPY_WITH_CPU_BASELINE_N > 0
```
No need to conditionally define this

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -13,6 +13,39 @@ static unsigned char npy__cpu_have[NPY_CPU_FEATURE_MAX];
+*/
+#if !defined(NPY_DISABLE_OPTIMIZATION) && NPY_WITH_CPU_DISPATCH_N > 0
+static int
+npy__cpu_dispatch_fid(const char *feature);
```
and this

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +132,132 @@ npy_cpu_dispatch_list(void)
+static int
+npy__cpu_baseline_fid(const char *feature)
+{
+    NPY_WITH_CPU_BASELINE_CALL(NPY__CPU_FEATURE_ID_CB, feature)
```
Conditionally use this line:
```
#if !defined(NPY_DISABLE_OPTIMIZATION) && NPY_WITH_CPU_BASELINE_N > 0
NPY_WITH_CPU_BASELINE_CALL(NPY__CPU_FEATURE_ID_CB, feature)
#endif
```

Then you could remove all the other occurances of `#if !defined(NPY_DISABLE_OPTIMIZATION) && NPY_WITH_CPU_BASELINE_N > 0`, like the one on line 243+250,

seiko2plus replied:
done

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +132,132 @@ npy_cpu_dispatch_list(void)
+static int
+npy__cpu_dispatch_fid(const char *feature)
+{
+    NPY_WITH_CPU_DISPATCH_CALL(NPY__CPU_FEATURE_ID_CB, feature)
```
as above, put the condition around this line

mattip reviewed (commented):
I suggest simplifying the code, the compiler will notice that the functions are constant and inline them.

seiko2plus commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +112,135 @@ npy_cpu_dispatch_list(void)
+    memcpy(disable_features, disenv, var_len);
+
+    //comma and space including (htab, vtab, CR, LF, FF)
+    const char *delim = ", \t\v\r\n\f";
```
I added extra separators, comma, and ascii2 special characters to make it more friendly with POSIX

seiko2plus commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +112,135 @@ npy_cpu_dispatch_list(void)
+    if (baseline_failure[0] != '\0') {
+        *(fptr-1) = '\0'; // trim the last space
+        PyErr_Format(PyExc_RuntimeError,
+            "The NumPy library was compiled with a minimal set of required optimizations: \n"
```
@mattip, also improved error messages.

seiko2plus commented:
@mattip, I made extra improvements to error messages and increasing the readability as much as possible.

anirudh2290 commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +112,135 @@ npy_cpu_dispatch_list(void)
+        if (!npy__cpu_have[feature_id]) {
+            PyErr_Format(PyExc_RuntimeError,
+                NPY__CPU_ENV_ERROR_HEAD
+                "CPU feature '%s' is not supported by your machine.\n",
```
if a non CPU supported feature is disabled, would it be better to warn than error in this case ?

mattip replied:
agreeed

seiko2plus replied:
agreed and done

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -13,6 +13,19 @@ static unsigned char npy__cpu_have[NPY_CPU_FEATURE_MAX];
 static void
 npy__cpu_init_features(void);
+/*
+ * Try to disable CPU dispatched features in runtime,
+ * if environment variable 'NPY_DISABLE_CPU_FEATURES' is defined,
+ * and it only use 'space' as seprator between features
```
```suggestion
 * Disable CPU dispatched features at runtime if environment variable
 * 'NPY_DISABLE_CPU_FEATURES' is defined.
 * Multiple features can be present, and separated by space, comma, or tab.
 * Raises an error if parsing fails or if the feature was not enabled
```

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -13,6 +13,19 @@ static unsigned char npy__cpu_have[NPY_CPU_FEATURE_MAX];
+*/
+static void
+npy__cpu_try_disable_env(void);
+/*
+ * check if NumPy build has CPU baseline features
+ * that aren't supported by the current hardware
+*/
```
Add a blank line to separate the declarations, reword docstring

```suggestion 

/* Ensure the build's CPU baseline features are supported at runtime */
```

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +112,135 @@ npy_cpu_dispatch_list(void)
+    if (baseline_failure[0] != '\0') {
+        *(fptr-1) = '\0'; // trim the last space
+        PyErr_Format(PyExc_RuntimeError,
+            "The NumPy library was compiled with a minimal set of required optimizations: \n"
+            "(" NPY_WITH_CPU_BASELINE ") but your machine doesn't support:\n(%s).\n",
```
Reword to use "baseline" since that is the term used to describe these features

```suggestion
            "NumPy was built with baseline optimizations: \n"
            "(" NPY_WITH_CPU_BASELINE ") but your machine doesn't support:\n(%s).\n",
```

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +112,135 @@ npy_cpu_dispatch_list(void)
+    #define NPY__MAX_VAR_LEN 1024 // More than enough for this era
+    size_t var_len = strlen(disenv) + 1;
+    if (var_len > NPY__MAX_VAR_LEN) {
+        PyErr_Format(PyExc_RuntimeError,
+            "Maximum acceptable length of NumPy's environment variable "
+            "'NPY_DISABLE_CPU_FEATURES' is %d",
+            NPY__MAX_VAR_LEN - 1
```
Shorten

```suggestion
        PyErr_Format(PyExc_RuntimeError,
            "Length of environment variable 'NPY_DISABLE_CPU_FEATURES' is %d, only %d accepted",
            var_len, NPY__MAX_VAR_LEN - 1 
```

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +112,135 @@ npy_cpu_dispatch_list(void)
+        if (npy__cpu_baseline_fid(feature) > 0) {
+            PyErr_Format(PyExc_RuntimeError,
+                NPY__CPU_ENV_ERROR_HEAD
+                "CPU feature '%s' is part of the minimal set of required optimizations:\n"
```
```suggestion
                "CPU feature '%s' is part of the baseline optimizations and cannot be disabled:\n"
```

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +112,135 @@ npy_cpu_dispatch_list(void)
+            PyErr_Format(PyExc_RuntimeError,
+                NPY__CPU_ENV_ERROR_HEAD
+                "CPU feature '%s' is not part of dispatched set of additional optimizations:\n"
+                "(" NPY_WITH_CPU_DISPATCH ").\n",
```
This should be a warning

mattip commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +112,135 @@ npy_cpu_dispatch_list(void)
+        feature = strtok(NULL, delim);
+    }
+#else
+    PyErr_Format(PyExc_RuntimeError,
```
also should be a warning

mattip commented on numpy/core/src/common/npy_cpu_features.h:
```diff
@@ -87,8 +87,20 @@ enum npy_cpu_features
 /*
  * Initialize CPU features
+ *
+ * This function
+ *  - detect CPU features avalibilty in runtime
+ *  - check the sanity of CPU baseline features
+ *  - disable CPU dispatched features in runtime,
+ *    depend on given values of environment variable 'NPY_DISABLE_CPU_FEATURES'
+ *
+ * It also trigger Py runtime error when
+ *  - NumPy build has CPU baseline features
+ *    that aren't supported by the current hardware
+ *  - The misuse of environment variable 'NPY_DISABLE_CPU_FEATURES'
+ *
```
```suggestion
 * This function
 *  - detects runtime CPU features
 *  - check that baseline CPU features are present
 *  - uses 'NPY_DISABLE_CPU_FEATURES' to disable dispatchable features
 *
 * It will set a RuntimeError when 
 *  - CPU baseline features from the build are not supported at runtime
 *  - 'NPY_DISABLE_CPU_FEATURES' tries to disable a baseline feature
 * and will warn if 'NPY_DISABLE_CPU_FEATURES' tries to disable a feature that
 * is not disabled (the machine or build does not support it, or the project was
 * not built with any feature optimization support)
```

seiko2plus committed:
ENH: several improvements to the CPU features detection

 - improve C comments
 - improve error messages
 - raise python warn rather then error if the disabled feature
   is not part of dispatched features or not supported by the running
   machine

Co-authored-by: Matti Picus <matti.picus@gmail.com>

seiko2plus commented on numpy/core/src/common/npy_cpu_features.c.src:
```diff
@@ -94,6 +111,163 @@ npy_cpu_dispatch_list(void)
+        return;
+    }
+    #define NPY__CPU_ENV_ERR_HEAD \
+        "During parsing environment variable 'NPY_DISABLE_CPU_FEATURES':\n"
```
@mattip, sorry but I changed again the format of error/warn messages. please check it again. thank you.

anirudh2290 reviewed (approved):
Tested this on my machine with a few different configs for the env variable, and it looks good to me !

mattip commented:
Thanks @seiko2plus
